gk/fib.c

```diff
@@ -1664,3 +1666,329 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+	while (index >= 0) {
+		memset(&dentry, 0, sizeof(dentry));
+		dentry.addr.proto = ETHER_TYPE_IPv6;
+		rte_memcpy(&dentry.addr.ip.v6, re6.ip,
+			sizeof(dentry.addr.ip.v6));
+		dentry.len = re6.depth;
+		fib = &ltbl->fib_tbl6[re6.next_hop];
+		ret = fillup_gk_fib_dump_entry(&dentry, fib);
+		if (ret < 0)
+			goto out;
+
+		lua_getglobal(l, handler);
+		lua_pushlightuserdata(l, &dentry);
+		if (arg == NULL)
+			lua_pushnil(l);
+		else
+			lua_pushlightuserdata(l, arg);
+
+		if (lua_pcall(l, 2, 1, 0) != 0) {
+			lua_error(l);
+			rte_free(dentry.ether_tbl);
+			ret = -1;
+			goto out;
+		}
+
+		rte_free(dentry.ether_tbl);
+
+		index = rte_lpm6_rule_iterate(&state6, &re6);
+	}
+
+out:
+	return ret;
+}
+
+#define CTYPE_STRUCT_GK_CONFIG_PTR "struct gk_config *"
+
+static int
+l_list_fib_entries(lua_State *l)
+{
+	int ret = 0;
+	uint32_t correct_ctypeid;
+
+	uint32_t ctypeid;
+	lua_Integer lpm_cfg;
+	struct gk_config *gk_conf;
+	void *arg;
+	struct gk_lpm *ltbl;
+
+	size_t handler_len;
+	const char *handler_c;
+	char fib_handler[128];
+	memset(fib_handler, 0, sizeof(fib_handler));
+
+	correct_ctypeid = luaL_get_ctypeid(l, CTYPE_STRUCT_GK_CONFIG_PTR);
+
+	/* First argument must be of type CTYPE_STRUCT_GK_CONFIG_PTR. */
+	luaL_checkcdata(l, 1, &ctypeid, CTYPE_STRUCT_GK_CONFIG_PTR);
+	if (ctypeid != correct_ctypeid)
+		luaL_error(l, "Expected `%s' as first argument",
+			CTYPE_STRUCT_GK_CONFIG_PTR);
+
+	/* Second argument must be a number. */
+	luaL_checktype(l, 2, LUA_TNUMBER);
+
+	/* Third argument must be a string. */
+	luaL_checktype(l, 3, LUA_TSTRING);
+
+	gk_conf = *(struct gk_config **)
+		luaL_checkcdata(l, 1, &ctypeid, CTYPE_STRUCT_GK_CONFIG_PTR);
+	lpm_cfg = lua_tointeger(l, 2);
+	handler_c = luaL_checklstring(l, 3, &handler_len);
+	if (handler_len > sizeof(fib_handler))
+		return 0;
+	strncpy(fib_handler, handler_c, handler_len);
+	if (lua_type(l, 4) == LUA_TNIL)
+		arg = NULL;
+	else
+		arg = lua_touserdata(l, 4);
+
+	ltbl = &gk_conf->lpm_tbl;
+
+	rte_spinlock_lock_tm(&gk_conf->lpm_tbl.lock);
+
+	if (lpm_cfg & LIST_IPV4_LPM_TABLE) {
+		ret = list_ipv4_fib_entries(l, ltbl, fib_handler, arg);
+		if (ret < 0)
+			goto out;
+	}
+
+	if (lpm_cfg & LIST_IPV6_LPM_TABLE)
+		ret = list_ipv6_fib_entries(l, ltbl, fib_handler, arg);
+
+out:
+	rte_spinlock_unlock_tm(&gk_conf->lpm_tbl.lock);
+
+	return ret;
+}
+
+static const struct luaL_reg dylib [] = {
+	{"list_fib_entries", l_list_fib_entries},
+	{NULL,               NULL}	/* Sentinel. */
+};
+
+const luaL_Reg *
+get_fib_dump_reg(void)
+{
+	return dylib;
+}
+
+struct ipaddr *
+get_fib_dump_entry_ip(struct gk_fib_dump_entry *entry)
+{
+	return &entry->addr;
+}
+
+struct ipaddr *
+get_fib_dump_entry_grantor_ip(struct gk_fib_dump_entry *entry)
+{
+	return &entry->grantor_ip;
+}
+
+struct ipaddr *get_fib_ether_dump_entry_nexthop(
+	struct gk_fib_ether_dump_entry *entry)
+{
+	return &entry->nexthop_ip;
+}
+
+struct ether_addr *get_fib_ether_dump_entry_mac(
+	struct gk_fib_ether_dump_entry *entry)
+{
```

Comment:
Since `struct gk_fib_ether_dump_entry` is already exported to Lua, why do we need `get_fib_ether_dump_entry_mac()` and its siblings?

Reply:
Since ``convert_ip_to_str()`` needs a pointer to the fields of ``struct gk_fib_ether_dump_entry`` with type of ``struct ipaddr``, we have to get the address of the its fields.

Reply:
Isn't there a way to get the address of field `d_addr` in Lua using FFI?

Reply:
I'm gessing that `dentry.d_addr` in Lua already is the address of field `d_addr` that can be passed to C.